Use threadlocals for caching of site root paths

ababic opened the pull request:
_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

ababic commented:
Regarding the current test failures: 

You'll see that most of them are related to general boosts in performance (query numbers drop, because cached values can be used in more places). However, there does seem to be something going on with hostnames being dropped from URLs where they were expected before, which I will dig some more into.

ababic commented:
Update: All the test failures are now query-number related only 🎉

ababic commented on wagtail/models/sites.py:
```diff
@@ -209,42 +212,58 @@ def get_site_root_paths():
+        cached for a thread's corresponding ``HttpRequest`` anyway.
+        """
+        global _root_paths_cache
+        _root_paths_cache = Local()
```
I was in two minds here about whether to trash the entire thing here, or just unset the value for the current thread only... Which felt 'fairer' from a 'multiple running instances' perspective, and probably has slightly less potential to cause issues should multiple threads be trying to access values concurrently. Thoughts?

ababic commented on wagtail/models/sites.py:
```diff
@@ -196,7 +199,7 @@ def clean_fields(self, exclude=None):
 
     @staticmethod
-    def get_site_root_paths():
+    def get_site_root_paths() -> List[SiteRootPath]:
```
Returning/caching a tuple feels more appropriate than a list here, but I would prefer to keep things backwards-compatible too

ababic commented on wagtail/models/sites.py:
```diff
@@ -15,6 +17,8 @@
 MATCH_DEFAULT = 2
 MATCH_HOSTNAME = 3
 
+_root_paths_cache = Local()
```
We could technically use this for other things, such as the request hostname/port and the current site (lazily determined from the former), but I think this name makes sense for now

ababic commented on wagtail/models/sites.py:
```diff
@@ -209,42 +212,58 @@ def get_site_root_paths():
+        result = getattr(_root_paths_cache, "value", None)
+        if result is not None:
+            # Return a copy to prevent accidental pollution of cached values
+            return deepcopy(result)
```
I guess another key benefit of the deepcopy here is that we prevent references to objects in the cache, 
which is important for enabling garbage collection / preventing memory leakage.

ababic commented:
I'm almost there with this! There are just a few api-related tests failing that I need to dig into.

ababic commented:
Okay, we're back to query number failures only, and I finally got my head around why certain tests were failing :)

A big change I've introduced this time is caching the sites themselves. The intention is that this cached data will be utilised by both `Page.get_site()` and `Site.find_for_request()` - extending the benefits of caching where full site data is needed.

zerolab commented:
I do love how the current failing tests are all on number of queries and they are all because fewer are executed 💪

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -16,33 +17,54 @@
 MATCH_HOSTNAME = 3
 
 
+class ClearableLocal(Local):
```
Suggestion: This should live elsewhere, perhaps in `utils`

ababic replied:
Yeah, I like that :)

ababic replied:
Resolved in [a50fcb9](https://github.com/wagtail/wagtail/pull/10018/commits/a50fcb9fbf5ce3077f1c5013d71117dedaa0a5c1)

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -213,54 +261,64 @@ def get_site_root_paths():
+        Convenience method for clearing both the `Site` and `SiteRootPath`
+        threadlocal caches.
+        """
+        _sites_cache.clear()
+        _site_root_paths_cache.clear()
```
Suggestion: Given we have additional methods for clearing the specific caches, I suggest we reuse those

ababic replied:
I must admit I'm on the fence with this one. I like the clarity here... Calling a manager method from the class feels a little awkward.

ababic replied:
I'm gonna keep this bit as it is unless anyone feels strongly about it?

RealOrangeOne commented on wagtail/signal_handlers.py:
```diff
@@ -18,13 +18,14 @@
-# Clear the wagtail_site_root_paths from the cache whenever Site records are updated.
 def post_save_site_signal_handler(instance, update_fields=None, **kwargs):
-    Site.clear_site_root_paths_cache()
+    # NOTE: applies to the current instance only
```
Question: Current instance of what?

ababic replied:
I'm open to ideas on the most appropriate language to use here, but what I mean here is the 'current running instance' - in situations where the project had multiple running instances (most likely in separate docker containers)

ababic replied:
Decided to remove these comments to avoid ambiguity. I think the docstrings on the methods themselves explain things well enough.

ababic replied:
Resolved in [601112e](https://github.com/wagtail/wagtail/pull/10018/commits/601112eebe944813e4abb1ed691f83af97062469)

RealOrangeOne commented on wagtail/tests/test_sites.py:
```diff
@@ -216,3 +217,140 @@ def test_result_order_when_multiple_sites_share_the_same_root_page(self):
+    def test_cache_clears_when_site_root_moves(self):
+        """
+        This tests for an issue where if a site root page was moved, all
+        the page urls in that site would change to None.
```
Question: This test doesn't seem to assert the URL

ababic replied:
Sorry, yes. Copied from the SiteRootPath tests without revising the docstring, but the steps are slightly different.

ababic replied:
These test were removed in f51287a in favour of extending TestSiteRootPathsCache to cover clearing of both caches

RealOrangeOne commented on wagtail/tests/tests.py:
```diff
@@ -443,38 +382,35 @@ def test_cache_clears_when_site_root_slug_changes(self):
         )
 
-        # Check url
-        self.assertEqual(translated_homepage.url, "/")
```
Question: Why are we no longer checking the URL?

ababic replied:
I've changed the nature of these tests subtly to be more unit-test like. The purpose of the test is now simply to assert that the cache is cleared. Whether or not the `url` method uses the cached values appropriately is a separate concern, and something to be tested elsewhere

ababic replied:
The docstrings for these tests have been simplified somewhat to set clearer expectations.

zerolab commented on wagtail/models/sites.py:
```diff
@@ -1,13 +1,14 @@
-from django.db.models import Case, IntegerField, Q, When
 from django.db.models.functions import Lower
 from django.http.request import split_domain_port
+from django.utils.local import ClearableLocal
```
```suggestion
from wagtail.utils.local import ClearableLocal
```

ababic replied:
Thanks @zerolab! That'll teach me to code while poorly!

zerolab replied:
I hear you! Did that so many times, and every time is a bit like "ah, I am not that poorly, I can do this" 🤦‍♂️

ababic commented:
Since my last update, I have:
1. Added the `WAGTAIL_PER_THREAD_SITE_CACHING` setting, that can be used to disable the cache in situations where it might cause confusion (e.g. tests)
2. Removed `Page. _get_site_root_paths()` and all of the tests that were focussed on the caching behaviour once offered there.
3. Updated `Page.get_site()` to take advantage of the cached data too. This method (indirectly) calls `Site.get_site_root_paths()`, triggering cache population, so the benefits are ripe for sharing here.
4. Reviewed and updated all of the failing tests. In most cases, I've leant more towards activating and populating the cache so (then adjusting the query counts accordingly), as I feel that's more reflective of reality, and has the added benefit of the tests themselves being more efficient. There's certainly no clear-cut indication of the best approach here, so fully appreciate that someone might have other ideas around this (please do share if so).

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -215,55 +254,75 @@ def get_site_root_paths():
+        this VM/container when changes that might affect the return values
+        are detected.
+
+        Any threads running in other VMs/containers will continue to use
+        outdated cached data for the remaining life of the thread. But,
+        any new requests/threads (or active threads that have not yet queried
+        the cache) should pick up the updated values.
```
Issue: This will result in different processes having different caches, leading to confusing and inconsistent behaviour. Processes may live for a very long time if there are no deployments, so it's important we not break functionality.

I think the only way to make this stable is to include some kind of TTL with the cache so processes get notified. Alternatively, something hierarchical, whereby the local cache has a very short TTL (a minute or so), and transparently hits a secondary cache that might be a little slower. Similar to L3 cache, RAM and disk behave.

ababic replied:
> This will result in different processes having different caches, leading to confusing and inconsistent behaviour.

I think you're ignoring the current state of play a little bit here. We already have different copies of the cached data everywhere, and we seem to be totally fine with that - That isn't the problem I'm solving here.

ababic replied:
> Processes may live for a very long time if there are no deployments, so it's important we not break functionality

Processes handled by the same container / VM will have their cached version of the data cleared should there be a change. Processes handled by other containers / VMs will not, but that isn't any worse than the current state of play.

ababic replied:
> transparently hits a secondary cache that might be a little slower.

Is this a suggestion to add cache backends into the mix on top of in-memory caching? That feels a tonne of extra complexity for not much gain here. From a pure system—design perspective, I get it. But realistically, the benefit would be minute compared solving the availability problem.

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -1,48 +1,60 @@
 MATCH_HOSTNAME = 3
 
 
+_sites_cache = ClearableLocal()
```
Issue: We want cache hits to be as high as possible, so we actually want the cache to be shared as much as possible, rather than "local". A global variable, so long as mutations were atomic, is probably closer to what we want here.

ababic replied:
@RealOrangeOne Let's remind ourselves about the current state of things:

1. The data is currently cached for the current `HttpRequest`
2. In order to access the data efficiently, the `HttpRequest` needs to be passed around everywhere, because thats where the references to cached data are stored
3. There is absolutely no invalidation of the cached data (it lives on for the duration of the request)

All I'm really changing here is where the data is stored...  putting it in a place where it can be accessed regardless of whether you have access to the `HttpRequest`. Each 'request' is handled by a separate thread in Django, so caching via `asgiref.local.Local` is equivalent to caching for the request.

ababic replied:
Putting the cached data in a central/known place also gives us the added benefit of being able to invalidate cached data, which is another massive bonus

ababic replied:
I would highly recommend taking a look at the code for Local here. It's a fantastic bit of code: https://github.com/django/asgiref/blob/main/asgiref/local.py

It's the same mechanism Django itself uses for caching the active/current language for the current user/request, without you needing access to the `HttpRequest`.

RealOrangeOne reviewed (commented):
Whilst I love the idea of making things much faster, sadly I have some fundamental issues.

ababic commented:
```
            /\             /\
           |`\\_,--="=--,_//`|
           \ ."  :'. .':  ". /
          ==)  _ :  '  : _  (==
            |>/O\   _   /O\<|
            | \-"~` _ `~"-/ |
           >|`===. \_/ .===`|<
     .-"-.   \==='  |  '===/   .-"-.
.---{'. '`}---\,  .-'-.  ,/---{.'. '}---.
 )  `"---"`     `~-===-~`     `"---"`  (
(               PROGRESS                )
 )              - over -               (
(              PURFECTION               )
'---------------------------------------'
```

jaap3 commented:
As far as I can tell `get_site_root_paths` is currently using the Django cache framework. So "some cache backend" is used, be it locmemcache, databasecache, rediscache etc. Most cache backends share the cache between processes, and can be invalidated from one process (except locmemcache). Besides that there's a 1 hour timeout on the cache value (probably for the locmemcache to catch up, if it was invalidated in another process). So, in most cases a call to `get_site_root_paths` will return a cached value (unless the cache is empty/has been invalidated/timed-out).

`wagtail/contrib/redirects/signal_handlers.py` seems to be caching something in `_page_urls_for_sites` on Page (or request objects), not sure why. This seems to be some hyper-specific optimisation for the handlers in there.

If I get you correctly, moving to threadlocals will require the cache to be filled on each request? I don't see how that's a benefit. Every request that uses `get_site_root_paths` will then have to make a database request to construct the list of root paths, instead of getting it from Django's cache.

If the overhead of accessing the cache is a concern the thread-locals approach could be layered on top. But ripping out the existing cache layer seems like a net negative.

jaap3 commented:
Regarding the reduced number of queries in tests, this can probably be (partially) explained by the use of the database cache backend in Wagtails test suite: 

https://github.com/wagtail/wagtail/blob/e407fb36127bca0b288aca36d4742b8906f1e8a2/wagtail/test/settings.py#L162-L171

I'm fairly certain that simply switching the backend to locmem for the test suite would also drop the number of queries.

ababic commented:
> As far as I can tell get_site_root_paths is currently using the Django cache framework. So "some cache backend" is used, be it locmemcache, databasecache, rediscache etc. Most cache backends share the cache between processes, and can be invalidated from one process (except locmemcache). Besides that there's a 1 hour timeout on the cache value (probably for the locmemcache to catch up, if it was invalidated in another process). So, in most cases a call to get_site_root_paths will return a cached value (unless the cache is empty/has been invalidated/timed-out).

All of this is true. I'm doing more than just caching site root paths here, though. I'm going to the root of the problem and caching the `Site` data from which they are derived. This way, that same data can be used for mapping requests and pages to sites without additional queries too.

> `wagtail/contrib/redirects/signal_handlers.py` seems to be caching something in _page_urls_for_sites on Page (or request objects), not sure why. This seems to be some hyper-specific optimisation for the handlers in there.

Yes, I haven't really alluded to this directly in any of my previous comments (because there's enough going on without muddying the waters), but a large part of wanting to make these changes stems from the weird things we currently do in places like that in order to benefit from caching. I don't think creating dummy/fake `HttpRequest`s to pass around should be necessary for that. I introduced `wagtail.coreutils.get_dummy_request()` a while ago to make this easier, but its still makes for a bit of a strange API.

> If I get you correctly, moving to threadlocals will require the cache to be filled on each request? I don't see how that's a benefit.

I would say the key benefits are:

1. It's easier to access the cached data. There are no longer any special requirements around needing access to an `HttpRequest` instance (or other object) that data has been cached for previously.
2. The caching is in-memory, which will always trump a lookup to external service for speed.

While it's true that using threadlocals has some overhead, I quite like that:

1. The data generated for each thread automatically expires when the thread does, without need for explicit invalidation
2. The still isn't shared between threads (doing that feels risky to me - at least, too risky for me to propose as an outside contributor).

> Every request that uses get_site_root_paths will then have to make a database request to construct the list of root paths, instead of getting it from Django's cache.

Sure, yep. There is one database query to fetch the site data for each request. However, the availability to that in-memory cached data makes a huge difference during the process of rendering a typical page with links to other pages in it:

- No need to always remember to use the `{% pageurl %}` tag in templates
- No need for separate cache/database lookups to turn page links in richtext values into fully-fledged URLs

> If the overhead of accessing the cache is a concern the thread-locals approach could be layered on top. But ripping out the existing cache layer seems like a net negative.

I wouldn't say it's a concern per se. More that once you have cached the `Site` data for the benefit of numerous other things, generating the site root path data becomes cheap enough that it doesn't really warrant the complication of talking to a caching service any longer.

ababic commented:
> Regarding the reduced number of queries in tests, this can probably be (partially) explained by the use of the database cache backend in Wagtails test suite:

Possibly, but I can't realistically go complicating this PR further by playing with test config. Perhaps one for a separate PR?

ababic commented:
I'd be happy to add a TTL to limit the potential for long-running processes of holding on to stale data, and also to revisit using a cache backend for better cross-request performance. 

If there's really no chance of the PR being accepted after that though, I'll close this and let someone else have a crack at this.

ababic commented:
Following a thoughtful and productive performance sub-team meeting, we agreed to make the following changes:

1. Revert TTL additions. 
2. Only clear cached data for the current thread when source data changes are detected via signals.
3. Restore the previous 'per request' caching behaviour for `Page.get_url()` so that we're less likely to break things (Also leaving the door open for optimisation when per-thread caching is disabled).

ababic commented:
@RealOrangeOne RE the long-running processes issue... would it be enough for them to set  `WAGTAIL_PER_THREAD_SITE_CACHING=False` and rely on passing objects around for caching?

ababic commented:
> I don't think so, we probably don't want them to need to remember to set this for background tasks, as if someone misses setting it, it could lead to some very strange behaviour. I think detecting it on their behalf is probably best.

Okay, cool. One to tackle in a follow-up PR :)

zerolab commented on wagtail/models/sites.py:
```diff
@@ -237,7 +237,7 @@ def clean_fields(self, exclude=None):
 
     @staticmethod
-    def get_site_root_paths(cls) -> Tuple[SiteRootPath]:
+    def get_site_root_paths(cls) -> tuple[SiteRootPath]:
```
Should this not be `Tuple`?

ababic commented:
@gasman All test issues are now resolved. As for the long-running threads... I feel advising people to regularly run `Site. clear_caches_for_thread()`  (say, at the start of management command runs) is about the best we can do.

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -68,12 +114,38 @@ def get_queryset(self):
     def get_by_natural_key(self, hostname, port):
         return self.get(hostname=hostname, port=port)
 
+    def get_all(self) -> Tuple["Site"]:
```
Question: Why is this a tuple? It ought to be a list.

ababic replied:
I found some information in the `weakref` [module docs](https://docs.python.org/3/library/weakref.html), which I think is relevant here:

> Several built-in types such as [list](https://docs.python.org/3/library/stdtypes.html#list) and [dict](https://docs.python.org/3/library/stdtypes.html#dict) do not directly support weak references but can add support through subclassing:

```
class Dict(dict):
    pass

obj = Dict(red=1, green=2, blue=3)   # this object is weak referenceable
```

> CPython implementation detail: Other built-in types such as [tuple](https://docs.python.org/3/library/stdtypes.html#tuple) and [int](https://docs.python.org/3/library/functions.html#int) do not support weak references even when subclassed.

Since [`asgiref.local` makes use of `threading.local`](https://github.com/django/asgiref/blob/main/asgiref/local.py#L76) and [`threading.local` makes use of `weakref`](https://github.com/python/cpython/blob/3.12/Lib/_threading_local.py#L183-L184), it feels like the safest and lowest complexity option here is to define and use custom list types, which I've now done.

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -215,55 +286,79 @@ def get_site_root_paths():
-    def clear_site_root_paths_cache():
-        cache.delete(SITE_ROOT_PATHS_CACHE_KEY, version=SITE_ROOT_PATHS_CACHE_VERSION)
+    def _get_cached_site_root_paths() -> Union[List[SiteRootPath], None]:
+        result = getattr(_site_root_paths_cache, "value", None)
+        if result is not None:
```
Nit: viva la walrus!

```suggestion
        if result := getattr(_site_root_paths_cache, "value", None):
```

ababic replied:
Beware suggesting use of walrus operators where the difference between `None` and other falsey values are important :)

laymonage replied:
They are not mutually exclusive, e.g.

```py
if result := getattr(_site_root_paths_cache, "value", None) is not None:
```

ababic replied:
@laymonage I'm not sure if this is on purpose or not, but I think that example would result in `result` ending up as a `bool` value, when what we really want is the `list` (empty or not) from the cache, if there is one.

laymonage replied:
Oops yep, you're right! It would need a parenthesis to work:

```py
if (result := getattr(_site_root_paths_cache, "value", None)) is not None:
```

at which point I think you're better off without the walrus 😄

RealOrangeOne commented on wagtail/models/sites.py:
```diff
@@ -215,55 +286,79 @@ def get_site_root_paths():
         - `root_url` - The scheme/domain name of the site (for example 'https://www.example.com/')
         - `language_code` - The language code of the site (for example 'en')
-        """
-        result = cache.get(
```
Issue: This will now always be recalculated every thread, rather than being able to rely on a shared cache (even if the cache is outside memory). It would be nice to maintain the existing cache as a form of tiered cache:

1. `asgiref.Local`
2. `django.core.cache`
3. Miss - recalculate

ababic replied:
Personally, I think that's a lot of extra code complexity and tests for not a lot of gain. But, I'm happy to be convinced otherwise. @gasman Do you have any thoughts on this?

ababic commented on wagtail/test/settings.py:
```diff
@@ -261,6 +261,12 @@
+# Disable per-thread caching of site/SiteRootPath data by default, as it tends to
+# cause confusion between tests that modify site data. It can be overridden on a
+# case-by-case basis when needed.
+WAGTAIL_PER_THREAD_SITE_CACHING = False
```
I think it's important to document this as something folks might want to set for their test suites to avoid any weirdness. I'm happy to document the setting itself, but it will definitely be worth a mention in the release notes, too.

ababic commented:
@RealOrangeOne / @gasman sorry for the direct mentions, but despite early enthusiasm, this PR seems to have been relegated to purgatory, despite having enormous value. I'd love your thoughts on what is needed to progress it.

Regarding the whole 'avoid this behaviour for management commands background tasks' aspect, I thought a tidy way of achieving that would be to add a simple Middleware that toggles the new behavior on. Yes, that makes it difficult to enable by default, but I'm okay with that.

ababic commented:
@kaedroho yes, I've thought a fair bit about that. Django serves each request in a new thread, so in 99% of cases, it's equivalent to caching data on the request object and reusing that data for the duration of the request (which is something we assume all projects are doing). So, the next request after site data had been updated would load the fresh data.

Where we need to exercise a bit more caution is where the data is cached outside the request/response cycle (where it becomes equivalent to a global cache) and where changes might need to be acted-upon off-the-back of the request where changes were made.

jams2 commented:
Hi @ababic , a couple of questions, and apologies if I've missed details from higher up the thread, it's pretty long :)

These changes here seem predicated on the idea that _Django serves each request in a new thread_. Isn't that something that's dependent on how you actually run the WSGI (or ASGI, I suppose) application? For instance, if running Gunicorn with sync workers, a worker process will handle multiple requests, sequentially. In that case, it seems that the lifetime of an asgiref Local will span multiple requests.

In that case, when the cache is cleared by the signal handler in this implementation, it seems that it would only be cleared in the process that's handling the request that initiated the signal, which would mean that there's the potential for an inconsistency in the cached site values between workers. The effect of which could be inconsistent page URLs being generated.

I'm also a little confused by the creation of the `WeakRefList` class, as there doesn't seem to be any creation of weak references.

ababic commented:
@jams2 if you'd like to better understand with how Locals works, you'll probably find the in-code documentation helpful: https://github.com/django/asgiref/blob/main/asgiref/local.py#L41

On the WeakRefList point... I'll dig into that when I have capacity to revise this. It's on pause for now whilst I focus on other things.

ababic commented:
For anyone else stumbling across this for the first time:
- Middleware will be added to enable this behaviour
- I already plan to scale things back a little by caching the SiteRootPath data only (not all site data)
